File: identity/node/src/face.ts
import { BaseClient, type FileInput } from './client.js'
import type {
  FaceDetectRequest,
  FaceDetectResponse,
  FaceMatchRequest,
  FaceMatchResponse,
} from './types.js'

export class FaceResource extends BaseClient {

  /**
   * Detect a face in an image and return bounding box + attributes (synchronous).
   *
   * @example
   * const result = await kernaq.face.detect({
   *   image: fs.createReadStream('selfie.jpg'),
   * })
   * console.log(result.detected, result.confidence)
   */
  detect(req: FaceDetectRequest): Promise<FaceDetectResponse> {
    return this.upload<FaceDetectResponse>('/face/detect', {}, [
      {
        field:       'image',
        value:       req.image as FileInput,
        filename:    req.imageName ?? 'image.jpg',
        contentType: 'image/jpeg',
      },
    ])
  }

  /**
   * Compare two face images and return a similarity score (synchronous).
   * Fields sent to the API: `face_a` and `face_b`.
   *
   * @example
   * const result = await kernaq.face.match({
   *   faceA: fs.createReadStream('id-photo.jpg'),
   *   faceB: fs.createReadStream('selfie.jpg'),
   * })
   * console.log(result.matched, result.confidence) // true, 99.2
   */
  match(req: FaceMatchRequest): Promise<FaceMatchResponse> {
    return this.upload<FaceMatchResponse>('/face/match', {}, [
      {
        field:       'face_a',
        value:       req.faceA as FileInput,
        filename:    req.faceAName ?? 'face_a.jpg',
        contentType: 'image/jpeg',
      },
      {
        field:       'face_b',
        value:       req.faceB as FileInput,
        filename:    req.faceBName ?? 'face_b.jpg',
        contentType: 'image/jpeg',
      },
    ])
  }
}

File: identity/go/kernaq/face.go
package kernaq

import "context"

// FaceResource handles stateless face detection and matching endpoints.
type FaceResource struct {
	c *client
}

// Detect detects the primary face in an image and returns attributes. Nothing stored.
func (r *FaceResource) Detect(ctx context.Context, req FaceDetectRequest) (*FaceDetectResponse, error) {
	imgName := nameOr(req.ImageName, "image.jpg")
	files := []filePart{{field: "image", reader: req.Image, filename: imgName, mime: mimeFromName(imgName)}}
	var out FaceDetectResponse
	if err := r.c.upload(ctx, "/face/detect", nil, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Match compares two face images. Returns matched + confidence. Nothing stored.
// ImageA is sent as "face_a", ImageB as "face_b".
func (r *FaceResource) Match(ctx context.Context, req FaceMatchRequest) (*FaceMatchResponse, error) {
	aName := nameOr(req.ImageAName, "face_a.jpg")
	bName := nameOr(req.ImageBName, "face_b.jpg")
	files := []filePart{
		{field: "face_a", reader: req.ImageA, filename: aName, mime: mimeFromName(aName)},
		{field: "face_b", reader: req.ImageB, filename: bName, mime: mimeFromName(bName)},
	}
	var out FaceMatchResponse
	if err := r.c.upload(ctx, "/face/match", nil, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
